doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,965 @@
+that features at least the following:
+
+- at least informal, predictable semantics
+- a typing system
+- standard programming facilities (i.e. datatypes)
+
+This document describes the implementation of such a language. The
+implementation of Ltac as of Coq 8.7 will be referred to as Ltac1.
+
+.. _ltac2_design:
+
+General design
+--------------
+
+There are various alternatives to Ltac1, such that Mtac or Rtac for instance.
+While those alternatives can be quite distinct from Ltac1, we designed
+Ltac2 to be closest as reasonably possible to Ltac1, while fixing the
+aforementioned defects.
+
+In particular, Ltac2 is:
+
+- a member of the ML family of languages, i.e.
+
+  * a call-by-value functional language
+  * with effects
+  * together with Hindley-Milner type system
+
+- a language featuring meta-programming facilities for the manipulation of
+  Coq-side terms
+- a language featuring notation facilities to help writing palatable scripts
+
+We describe more in details each point in the remainder of this document.
+
+ML component
+------------
+
+Overview
+~~~~~~~~
+
+Ltac2 is a member of the ML family of languages, in the sense that it is an
+effectful call-by-value functional language, with static typing à la
+Hindley-Milner. It is commonly accepted that ML constitutes a sweet spot in PL
+design, as it is relatively expressive while not being either too lax
+(contrarily to dynamic typing) nor too strict (contrarily to say, dependent
+types).
+
+The main goal of Ltac2 is to serve as a meta-language for Coq. As such, it
+naturally fits in the ML lineage, just as the historical ML was designed as
+the tactic language for the LCF prover. It can also be seen as a general-purpose
+language, by simply forgetting about the Coq-specific features.
+
+Sticking to a standard ML type system can be considered somewhat weak for a
+meta-language designed to manipulate Coq terms. In particular, there is no
+way to statically guarantee that a Coq term resulting from an Ltac2
+computation will be well-typed. This is actually a design choice, motivated
+by retro-compatibility with Ltac1. Instead, well-typedness is deferred to
+dynamic checks, allowing many primitive functions to fail whenever they are
+provided with an ill-typed term.
+
+The language is naturally effectful as it manipulates the global state of the
+proof engine. This allows to think of proof-modifying primitives as effects
+in a straightforward way. Semantically, proof manipulation lives in a monad,
+which allows to ensure that Ltac2 satisfies the same equations as a generic ML
+with unspecified effects would do, e.g. function reduction is substitution
+by a value.
+
+Type Syntax
+~~~~~~~~~~~
+
+At the level of terms, we simply elaborate on Ltac1 syntax, which is quite
+close to e.g. the one of OCaml. Types follow the simply-typed syntax of OCaml.
+
+.. productionlist:: coq
+   type             : ( `type`, ... , `type` ) `typeconst`
```

Comment:
Are the namespaces actually global?  Or intended to be global?  The `coq` after `productionlist` is meant to identify different sets of productions, which when I tried it seemed to correctly distinguish the different sets.

On the other hand, it will be very confusing for users to see multiple inconsistent productions for the same non-terminals in different chapters.  So I agree the names should be changed in doc.  Also, the grammar appears to repeat many productions that are already in Coq.

Reply:
They are global in the sense that when you refer to an existing token through `:token:` or `:n:` or `:cmd:`, etc., then you have no way to say to which productionlist it belongs. I still have no idea what is the exact purpose of the qualifier after `productionlist`.

Reply:
Defining the same nonterminal twice under "coq" gives an error; defining under separate labels does not.

But as you note, token doesn't work with this mechanism.

Reply:
> Defining the same nonterminal twice under "coq" gives an error; defining under separate labels does not.

@jfehrle then why didn't redefining `term` fail?

Reply:
> then why didn't redefining `term` fail?

No idea.  I have seen duplicate nonterminal messages in the past but apparently they aren't generated in all the cases we would like.  Maybe the duplicate checking doesn't work across multiple doc files.  The Sphinx doc doesn't go into such detail.  Or it may be a Sphinx bug.